ckan/cli/__init__.py

```diff
@@ -5,9 +5,60 @@
 import click
 import logging
 from logging.config import fileConfig as loggingFileConfig
+from configparser import ConfigParser
 
 log = logging.getLogger(__name__)
 
+class CKANConfigLoader(object):
+    def __init__(self, filename):
+        self.filename = filename = filename.strip()
+        self.parser = ConfigParser()
+        self.parser.read(filename)
+
+        defaults = {
+            'here': os.path.dirname(os.path.abspath(filename)),
+            '__file__': os.path.abspath(filename)
+            }
+        self._update_defaults(defaults)
+
+    def _update_defaults(self, new_defaults, overwrite=True):
+        for key, value in new_defaults.iteritems():
+            if not overwrite and key in self.parser._defaults:
+                continue
+            self.parser._defaults[key] = value
+
+    def get_context(self):
+        global_conf = self.parser.defaults().copy()
+        local_conf = {}
+        section = 'app:main'
+        options = self.parser.options(section)
+
+        for option in options:
+            if option in global_conf:
+                    continue
+            local_conf[option] = self.parser.get(section, option)
+
+        return CKANLoaderContext(global_conf, local_conf).config()
+
+class CKANLoaderContext(object):
+    def __init__(self, global_conf, local_conf):
+        self.global_conf = global_conf
+        self.local_conf = local_conf
+
+    def config(self):
+        conf = AttrDict(self.global_conf)
+        conf.update(self.local_conf)
+        conf.local_conf = self.local_conf
+        conf.global_conf = self.global_conf
+        conf.context = self
+        return conf
+
+class AttrDict(dict):
+    """
+    A dictionary that can be assigned to.
```

Comment:
@pdelboca can you please perhaps provide more info as to why we need this dict and why it's used? i.e. could you just use a local dict constructor within the `def config(self)` method and just call it conf; so something like:

```
def config(self):
        conf = dict(self.global_conf)
        conf.update(self.local_conf)
....
```
(- https://python-reference.readthedocs.io/en/latest/docs/functions/dict.html)

Reply:
The current protocol of the config object is a mix between object and dictionary: 
 - sometimes their attributes are access using `conf.local_conf` 
 - sometimes using `conf.get('ckan.site_url', 'http://0.0.0.0')`
 - sometimes they are separated into local_conf and global_conf and each one of those should be dictionaries that can be accessed with `global_conf['__file__']`

This `AttrDict` class implementation was taken from [Pylon's pastedeploy](https://github.com/Pylons/pastedeploy/blob/9cbc584a86bd1ae425bd0c0c3c00617d7e59475d/paste/deploy/loadwsgi.py#L726) and it respect the current implementation.

Another option is to drop everything and use a simple conf `dict`, with no `local_conf` nor `global_conf` and refactor every call to it in the code. Maybe someone in the @ckan/core can give some insights on what is expected.

Reply:
@pdelboca @amercader @Zharktas now it's clear, so we're wrapping it in an `object` to be able to assign to it. 

In my taste this is a bit wrong as composition should have been applied here to avoid ambiguity (but also to have readability :))  , so we should have a dict part that is explicitly accessible and an arbitrary namespace, explicitly accessible to set values into.

I propose we take this opportunity to clean up the usage of this, what do you think?

Reply:
@pdelboca did you confirm the the `create()` method is redundant btw (https://github.com/ckan/ckan/pull/5135#pullrequestreview-337696537) ?